### PR TITLE
Format haddocks attached to arguments of data constructors with pipe

### LINKED ISSUE #828: Inline Haddock comments on data fields are not formatted correctly
**Describe the bug**
The snippet
```haskell
data A = A {- | a number -} Int Bool
```
is formatted to
```haskell
data A = A Int
           -- ^ a number Bool
```
which is obviously incorrect (differing AST).

**Environment**
0.3.1.0 and e003e11a569f9801b2a3f0cc01a66c883d597a3d

**Additional context**
Found via #821.

Note that the snippet above did not parse with GHC < 9.0.

### LINKED ISSUE #844: Ormolu uses -- ^ style to annotate arguments of a data constructor
Ormolu currently transforms this snippet

```haskell
-- | Foo.
data Foo
  = Foo
      -- | Bar
      Char
      -- | Baz
      Int
```

into

```haskell
-- | Foo.
data Foo
  = Foo
      Char
      -- ^ Bar
      Int
      -- ^ Baz
```

which seems to be at odds with how we format Haddocks in other contexts. The version with pipes (`-- |`) is accepted by the latest `ghc-lib-parser` and so it makes sense to migrate to that style someday.

### PR DESCRIPTION
**Pull request by mrkkrp**
Close #844, close #828.

### CODE REVIEW

File: data/examples/declaration/data/unnamed-field-comment-2-out.hs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+-- | Describes what sort of dictionary to generate for type class instances
+data Evidence
+  = -- | An existing named instance
+    NamedInstance (Qualified Ident)
+  | -- | Computed instances
```
Comment (mrkkrp) [inline data/examples/declaration/data/unnamed-field-comment-2-out.hs:5]: This is a bit insane, but Agda authors do not use Haddock correctly, so too bad.

**Review by amesgen — COMMENTED** [review]
I got reminded of #828, and indeed, the snippet
```haskell
data A = A {- | a number -} Int Bool
```
from that issue is now formatted to
```haskell
data A = A
           -- | a number
           Int
           Bool
```
which has equal AST, but fails the idempotency check. Unless it turns out to be surprisingly difficult: What about adapting the logic here such that this snippet is idempotent? You are already adding some logic that depends on whether Haddocks are present.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,9 @@
+## Unreleased
+
+* Haddocks attached to arguments of a data constructor are now formatted in
+  the pipe style (rather than the caret style), consistent with everything
+  else. [Issue 844](https://github.com/tweag/ormolu/issues/844).
```
[the remark this answers is not in the document]
Comment (mrkkrp) [reply CHANGELOG.md]: I added a note about this :+1:

**Comment by mrkkrp** [comment]
Great catch re #828. I believe this should be fixed now.

File: src/Ormolu/Printer/Meat/Declaration/Data.hs [hunk 3]
```diff
@@ -264,5 +268,18 @@ isSingleConstRec _ = False
...
 hasHaddocks = any (f . unLoc)
   where
-    f ConDeclH98 {..} = isJust con_doc
+    f ConDeclH98 {..} =
+      isJust con_doc
+        || ( case con_args of
+               PrefixCon [] xs ->
+                 conArgsHaveHaddocks (hsScaledThing <$> xs)
+               _ -> False
+           )
```
Comment (amesgen) [inline src/Ormolu/Printer/Meat/Declaration/Data.hs:277]: Feel free to ignore:
```suggestion
      isJust con_doc || case con_args of
        PrefixCon [] xs ->
          conArgsHaveHaddocks (hsScaledThing <$> xs)
        _ -> False
```

**Review by amesgen — APPROVED** [review]
:100:

**Pull request merged**